lib/withOnyx.d.ts

```diff
@@ -0,0 +1,40 @@
+import {IsEqual} from 'type-fest';
+import {Key, CollectionKey, Selector, Value} from './types';
+
+type BaseMapping<TComponentProps, TOnyxProps> = {
+    canEvict?: boolean | ((props: Omit<TComponentProps, keyof TOnyxProps>) => boolean);
+    initWithStoredValues?: boolean;
+};
+
+type Mapping<TComponentProps, TOnyxProps, TOnyxProp extends keyof TOnyxProps, TOnyxKey extends Key | CollectionKey> = BaseMapping<TComponentProps, TOnyxProps> &
+    (
+        | (IsEqual<Value[TOnyxKey] | null, TOnyxProps[TOnyxProp]> extends true
+              ? {
+                    key: TOnyxKey | ((props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey);
+                }
+              : never)
+        | {
+              key: TOnyxKey;
+              selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
+          }
+        | {
+              key: (props: Omit<TComponentProps, keyof TOnyxProps>) => TOnyxKey;
+              selector: Selector<TOnyxKey, TOnyxProps[TOnyxProp]>;
+          }
+    );
+
+// TODO: Missing implementation.
+declare function createOnyxSelector<TOnyxKey extends Key | CollectionKey, TResultData>(selector: Selector<TOnyxKey, TResultData>): Selector<TOnyxKey, TResultData>;
```

Comment:
If it's possible I would omit such workarounds. It may be hard for devs to understand why and when it is needed.

Reply:
I didn't understand your point @blazejkustra , could you elaborate more? This utility function must have an implementation in Onyx, that's why I added the comment.

Reply:
Sorry @fabioh8010! What I mean is that I would avoid such helpers as `createOnyxSelector` at all cost. I would completely remove it from the `react-native-onyx` codebase. From the contributor perspective it'll be hard to tell how to use such utility and why it is actually needed only in specific cases. Especially that we have to manually pass type parameters to use it, right? @fabioh8010 

What I would do is either completely remove it, or use it for all selectors for consistency @fabioh8010

Reply:
I agree with @blazejkustra let's remove it unless we can someone incorporate it in the types itself

Reply:
> Especially that we have to manually pass type parameters to use it, right? @fabioh8010

Yes, we need to pass type parameters in order to work correctly.

---

@hayata-suenaga @blazejkustra The usage of `createOnyxSelector` would help mitigate a problem we have in both `Onyx prop with string collection key and selector` and `Onyx prop with function collection key and selector` cases:

> The function signature of `selector` is not being correctly inferred and will allow specifying a different type to `value`. Using `createOnyxSelector` can help mitigate this issue. ❌

But as @blazejkustra pointed out we still have to make sure `createOnyxSelector` has the right type parameters, otherwise we have the same issue.

With that said, let's remove it then!